### ISSUE TITLE
devel/viewvc and www/mod_dav_svn wipe each other out?

### ISSUE THREAD

**Issue by rwv37** [jrmarino/synth#174]
I used synth for the first time last night, after having used portmaster for many years. I did a "synth upgrade-system", which seemed like it was going great, but eventually I went to sleep while it was still running (devel/llvm80 was taking hours and hours).

When I woke up, it had completed, and I didn't notice any error messages. But then I stopped Apache and tried to start it; it failed to start because of some sort of problem with mod_dav_svn. I then did a "pkg version -v", and saw that mod_dav_svn had been uninstalled. So, that seemed kind of odd, but I figured I would take this as an opportunity to try my first "synth install" - "synth install www/mod_dav_svn".

That seemed to work fine, and then I started Apache (successfully). Checked that Subversion was working via Apache, which it was. Then I checked it through viewvc. Unfortunately, my browser was given an error message saying that I did not have permission to access viewvc. So, I did a "pkg version -v" again, and noticed that viewvc had been uninstalled. I then tried my second "synth install" - "synth install dev/viewvc".

The synth install seemed to work fine. I then stopped Apache, tried to restart it, and got the error about mod_dav_svn again. Sure enough, "pkg_version -v" revealed that mod_dav_svn had again been uninstalled.

So now I tried my first "synth install" having multiple ports - "synth install www/mod_dav_svn devel/viewvc". Again the synth install seemed to work fine, but mod_dav_svn was still not installed.

Then I tried installing the two packages in the other order - "synth install devel/viewvc www/mod_dav_svn". The result was that mod_dav_svn was installed, but viewvc was uninstalled.

I then looked at the output of that last synth install a little closer, and noticed:

```
Checking integrity... done (2 conflicting)
  - subversion-1.14.0 [Synth] conflicts with subversion-lts-1.10.6 [installed] on /usr/local/bin/fsfs-access-map
  - subversion-1.14.0 [Synth] conflicts with subversion-lts-1.10.6 [Synth] on /usr/local/bin/fsfs-access-map
Checking integrity... done (0 conflicting)
The following 5 package(s) will be affected (of 0 checked):

Installed packages to be REMOVED:
        py27-subversion-lts: 1.10.6
        subversion-lts: 1.10.6
        viewvc: 1.2.1

New packages to be INSTALLED:
        mod_dav_svn: 1.14.0 [Synth]
        subversion: 1.14.0 [Synth]

Number of packages to be removed: 3
Number of packages to be installed: 2
```

I am pretty confident that I never (explicitly) asked for "subversion-lts" to be installed (I mean ever, not just during all of this); I've always just used the subversion port. The "py27-subversion-lts" thing was surprising too - I don't remember ever having seen it before.

I know that there have been recent EOL warnings about the various py27 things, so I'm guessing that may have something to do with what's happening here?

I always read /usr/ports/UPDATING religiously before updating, and I didn't remember seeing anything recent about subversion or mod_dav_svn or viewvc. After all of this, I went back and read it again, and still didn't see anything.

The last time I performed an upgrade to my installed ports was a week ago (using portmaster). It worked fine.

Any idea on what's going wrong here, or how I can get this all working again? Thanks.

**Comment by rwv37** [comment]
Oh, I don't know whether this is relevant or not, so I should probably mention:

Immediately after installing synth, and running it for the very first time (with a "synth status"), it actually failed with a "REPLICANT.SCENARIO_UNEXPECTED" error. I googled around a bit, and found [this page](https://github.com/jrmarino/synth/issues/145). I then tried "/usr/sbin/pwd_mkdb -p /etc/master.passwd" as suggested there by zamzamdesu, and that at least seemingly resolved the REPLICANT.SCENARIO_UNEXPECTED error. Only after that and another "synth status" did I do the initial "synth upgrade-system" as mentioned in my original comment here.

I should also probably note that I did that pretty blindly; I did not and do not understand the explanations in that thread of what all this means or why the suggestion would help.

**Comment by jrmarino** [comment]
"upgrade system" is just a convenience command to tell pkg(8) to use a local repository to upgrade the system.  Whatever happens as a result of mixed multiple repositories (freebsd official + local) combined with local building and install via portmaster is behavior dictated by pkg(8).

To put it another way, you could have limited synth to just build packages and generated a respository ... then separately executed pkg(8) to use that repository and the same thing would have happened.  So it's basically out of scope of synth.  It's going to require you to understand how pkg(8) works a little better.

Regarding the pwd_mkdb solution, that sounds like the correct solution so that's not a problem.

**Comment by rwv37** [comment]
Thanks, but I don't fully understand what you're saying.

To back up a moment, here's what I have done for years and years: I build everything from ports, no packages (except maybe once every few years where I just can't get something to compile). Every week or so, I do a "portsnap fetch update", I read the latest posts in /usr/ports/UPDATING and do whatever applicable things are listed there, and then I do  a "portmaster -advw" to upgrade all installed ports to their latest version.

What I was hoping for out of synth, and what I thought I was going to get out of "synth upgrade-system", was essentially a drop-in replacement for that "portmaster -advw" step having some additional benefits, especially multiple simultaneous compilations.

So, forgetting about what I *did*, is what I was *hoping for* something that synth is appropriate for?

If so, what *should I have done* in order to accomplish what I was hoping for? And what *did I do wrong* that caused, for example, devel/subversion to be replaced by devel/subversion-lts, which I didn't explicitly ask for and which I've never seen happen before when I used portmaster for years and years?

**Comment by jrmarino** [comment]
okay, so the reason you don't fully understand is that there are 2 worlds.
You are from the "old world" where builds were done live and "dirty".  That's how portmaster was designed and that's how it was used.  If dependencies that were installed were old, it didn't matter, portmaster did the best it could to build it and install it.  There was no coherence to the current state of the ports tree.  It's from an era where it was recommended to build every port, then install.

The "new world" was introduced by poudriere.  That's where the builds are clean and reproducible.  Dependencies are brought up to date before building the target ports.  The focus moved to binary packages where building your own first became optional, then became discouraged.  After the build, the now up-to-packages are put into a cohesive collection called a "repository".  The pkg(8) program works on a repository basis, replacing all out-of-date software during the upgrade process (unless you force it not to which often leaves the system in a weird state, not recommended).

So there is no "drop-in" replacement for postmaster and portmaster's work flow.  It's a workflow that's been criticized but it has die-hard fans (often those not exposed to any other way).

Synth is basically similar to poudriere, but it supposed to be easier to learn and be productive with.  If you are a poudriere user, the benefits synth provides over poudriere may not be great enough to change (and nobody recommended to switch).  From the feedback I've gotten, synth was successful in this goal.

As for what you "did wrong", you didn't do anything wrong other than mixing live builds with repository builds.  You should switch to completely doing repository builds.  That is, either build everything with poudriere, or build everything with synth, or install everything from official packages.  (synth has the ability to mix official packages with custom local built ones but that's an intermediate topic).  Once you have all the packages from a single source, then updating will lead to a predictable experience.  That weirdness you're seeing is portmaster patched to use pkg(8) (which it far pre-dates pkg(8) and the original pkg-tools support was removed) and it's no enforcing cohesive installations.

I'd say pick a method, reinstall all packages from that method so you have a good starting point, then keep using it.  Everything will be predictable and reliable then.

I hope that sort of makes sense ...

**Comment by rwv37** [comment]
OK, great, yes, that does make sense (I think). Thanks. And I'm sorry to keep bothering you about this, but just to make sure that I really am understanding, is this correct:

What I should be doing now on my machine, if I want to switch from portmaster to synth, is (A) uninstall all the ports I've got installed; (B) reinstall the ports that I want using synth; (C) forget about portmaster from now on.

**Comment by jrmarino** [comment]
well, the order would be:
1) gather a list of top level ports you want installed (e.g. desktop, libreoffice, whatever)
2) put those in a file
3) send those to synth to build
4) create a repository
5) uninstall all packages
6) Tell pkg(8) to install those top-level packages you listed in step 2
7) from then on, use that list to keep your packages up to date, appending it as you add software.

**Comment by jrmarino** [comment]
and don't conflate synth and pkg.  synth builds the packages and creates the repository.  pkg keeps packages on your system up to date based on the repository it's connected with.

**Comment by rwv37** [comment]
OK, thanks again, but now I'm a bit confused again. What I was imagining was the following, and I don't understand why I shouldn't do it this way:

(1) Uninstall everything first
(2) "make install clean" synth
(3) "synth install" the top level ports I want

... and from then on just always use "synth install" to install new things, and "synth upgrade-system" to get existing things up to date.

Are you suggesting that just because it might be convenient for me to have a GUI (for example) installed while I'm doing all this? As opposed to suggesting it because what I was thinking is actively incorrect in some way?

**Comment by jrmarino** [comment]
you should create the repository before you remove any packages.
And I warned you about conflating synth and pkg.  "synth install" is just synth running pkg for you.  You can run them separately. 

The problem with "upgrade system" is that it doesn't check if the repository is in a good state.  maybe there was a blocked or failed port.  You don't want to install packages until the all the packages are built.  Honestly I feel that there is such a potential for unwanted behavior that I never should have had synth command pkg because many newbies don't understand what they are doing.  And they try to schedule package build via cron.  it works for a while but something inevitably breaks.

Until you reach expect level, I would suggest you don't use "synth install" and "synth upgrade-(system)".  Get into the habit of creating repositories and then using pkg(8) directory.  No, it's not nearly as convenient and yes, it might be overkill because probably pkg(8) will do the right thing with the local packages.   Okay, let's compromise: you can use "synth install"  and "synth upgrade", but avoid "synth upgrade-system" for a while.

Anyway, I put it in that order to make sure you had a full set of packages *before* you removed everything from your system (including synth itself).

**Comment by jrmarino** [comment]
also the process of removing everything and then reinstalling top level ports is what I call "bulletproof".   It always works and puts the system in a good config.  It can be done anytime you feel the system state is wrong and maybe not fixable.

**Comment by rwv37** [comment]
OK, I'll do that. Thanks again. For what it's worth, I think a great deal of my confusion (both original and current) was largely based upon the documentation on the main page of this git repository:

(1) The very first thing it says:

"Synth is an advanced concurrent (parallel) ports building tool aimed at regular users that prefer or require building their own packages from source. Synth will build packages in a clean environment that exactly mirrors the system it builds on, creates a local repository, installs a pkg repository configuration file that causes the local packages to be used with the highest priority, and automatically upgrades the system with a single command."

"... aimed at regular users...", but apparently it's something that shouldn't be used by noobs until they're expert level

"... exactly mirrors the system it builds on...", except that it doesn't (e.g. the original weirdness with mod_dav_svn, viewvc, subversion, and subversion-lts)

"... automatically upgrades the system with a single command", except you shouldn't do that unless you're an expert

(2) "And I warned you about conflating synth and pkg. "synth install" is just synth running pkg for you."

The docs say it's the same as "synth build" except it doesn't ask you questions. I understand that it runs pkg. I don't understand that it's *just* running package.

(3) "Okay, let's compromise: you can use "synth install" and "synth upgrade", but avoid "synth upgrade-system" for a while."

I don't see anything in the docs mentioning "synth upgrade". Just "synth upgrade-system".

Anyway, thanks again.

**Comment by jrmarino** [comment]
the "mirror" comment means it uses a clean version of the system (what it might look like freshly installed off a cd).  E.g. it might build differently on a system with a different cpu.  It does NOT refer to packages that installed on that system.

i might have forgotten the terms and nuances.  I haven't used or developed synth beyond bug fixing in many years.

I would stand by the "regular users" comment.  I just gave them a loaded gun with "synth upgrade-system" that too many people have shot their foot off because they didn't fully understand what could happen.  I am probably being too conservative.
